google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -181,9 +181,419 @@ class ResumableAsyncStreamingReadWriteRpc {
+          });
+        case State::kInitialized:
+          read_future = stream_->Read();
+          assert(!in_progress_read_.has_value());
+          in_progress_read_.emplace();
+      }
+    }
+
+    return read_future.then(
+        [this](future<absl::optional<ResponseType>> optional_response_future) {
+          return OnReadFutureFinish(optional_response_future.get());
+        });
+  }
+
+  future<bool> Write(RequestType const& r, grpc::WriteOptions o) override {
+    future<bool> write_future;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      switch (stream_state_) {
+        case State::kShutdown:
+          return make_ready_future(false);
+        case State::kRetrying:
+          assert(!write_reinit_done_.has_value());
+          write_reinit_done_.emplace();
+          return write_reinit_done_->get_future().then(
+              [](future<void>) { return make_ready_future(false); });
+        case State::kInitialized:
+          write_future = stream_->Write(r, o);
+          assert(!in_progress_write_.has_value());
+          in_progress_write_.emplace();
+      }
+    }
+
+    return write_future.then([this](future<bool> write_fu) {
+      return OnWriteFutureFinish(write_fu.get());
+    });
+  }
+
+  future<void> Shutdown() override {
+    promise<void> root_promise;
+    future<void> root_future = root_promise.get_future();
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      switch (stream_state_) {
+        case State::kShutdown:
+          return make_ready_future();
+        case State::kRetrying:
+          stream_state_ = State::kShutdown;
+          root_future = root_future.then(ChainFuture(
+              retry_promise_->get_future().then([this](future<void>) {
+                std::lock_guard<std::mutex> g{mu_};
+                CompleteStreamLockHeld(Status());
+              })));
+          break;
+        case State::kInitialized:
+          stream_state_ = State::kShutdown;
+          if (in_progress_read_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_read_->get_future()));
+          }
+          if (in_progress_write_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_write_->get_future()));
+          }
+          std::shared_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+              stream = std::move(stream_);
+          CompleteStreamLockHeld(Status());
+          root_future = root_future.then([stream](future<void>) {
+            return future<void>(stream->Finish());
+          });
+      }
+    }
+    root_promise.set_value();
+    return root_future;
+  }
+
+ private:
+  using UnderlyingStream =
+      std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>;
+
+  enum class State { kRetrying, kInitialized, kShutdown };
+
+  future<absl::optional<ResponseType>> OnReadFutureFinish(
+      absl::optional<ResponseType> optional_response) {
+    promise<void> in_progress_read;
```

Comment:
You can use:

```
auto in_progress_read = promise<void>(null_promise_t{});
```

to initialize a promise that is "empty", and save yourself some initialization time.